packages/core/core/src/RequestTracker.js

```diff
@@ -1108,6 +1117,74 @@ export default class RequestTracker {
     return {api, subRequestContentKeys};
   }
 
+  generateInvalidationReport(): void {
+    let nodes = {};
+    let invalidationRelations = {};
+    for (let {invalidated, cause} of this.graph.invalidationReport) {
+      let invalidatedNode = this.graph.getNode(invalidated);
+      let causeNode =
+        cause !== undefined ? this.graph.getNode(cause) : undefined;
+      if (
+        invalidatedNode === null ||
+        causeNode === null ||
+        invalidatedNode === undefined ||
+        causeNode === undefined
+      ) {
+        continue;
+      }
+
+      if (invalidatedNode.id !== 'Main') {
+        if (!invalidationRelations[invalidated]) {
+          invalidationRelations[invalidated] = [];
+        }
+
+        if (!invalidationRelations[invalidated].includes(cause)) {
+          invariant(cause !== undefined);
+          invalidationRelations[invalidated].push(cause);
+        }
+      }
+
+      invariant(invalidatedNode.type === 1);
+      nodes[invalidated] = {
+        id: invalidatedNode.id,
+        type: invalidatedNode.type,
+        requestType: invalidatedNode.requestType,
+        invalidateReason: invalidatedNode.invalidateReason,
+      };
+
+      invariant(cause !== undefined);
+      if (
+        causeNode.requestType !== undefined &&
+        causeNode.invalidateReason !== undefined
+      ) {
+        nodes[cause] = {
+          id: causeNode.id,
+          type: causeNode.type,
+          requestType: causeNode.requestType,
+          invalidateReason: causeNode.invalidateReason,
+        };
+      } else {
+        nodes[cause] = {
+          id: causeNode.id,
+          type: causeNode.type,
+        };
+      }
+    }
+
+    const fs = require('fs');
+    fs.writeFile(
+      'cache-invalidation-report.json',
+      JSON.stringify(
+        {invalidationRelations: invalidationRelations, nodes: nodes},
+        undefined,
+        4,
+      ),
+      function (err) {
```

Comment:
This is running async via a callback but we're not waiting for completion anywhere. We should be awaiting a promise here instead. If you migrate to `outputFS` then this should be easy.